copulas/bivariate/base.py

```diff
@@ -0,0 +1,159 @@
+from enum import Enum
+
+import numpy as np
+from scipy import stats
+
+
+class CopulaTypes(Enum):
+    CLAYTON = 0
+    FRANK = 1
+    GUMBEL = 2
+
+
+class Bivariate(object):
+    """Base class for all bivariate copulas."""
+
+    copula_type = None
+    _subclasses = []
+
+    @classmethod
+    def _get_subclasses(cls):
+        subclasses = []
+        for subclass in cls.__subclasses__():
+            subclasses.append(subclass)
+            subclasses.extend(subclass._get_subclasses())
+
+        return subclasses
+
+    @classmethod
+    def subclasses(cls):
+        if not cls._subclasses:
+            cls._subclasses = cls._get_subclasses()
+
+        return cls._subclasses
+
+    def __new__(cls, copula_type=None):
+        if not isinstance(copula_type, CopulaTypes):
+            if (isinstance(copula_type, str) and copula_type.upper() in CopulaTypes.__members__):
+                copula_type = CopulaTypes[copula_type.upper()]
+            else:
+                raise ValueError('Invalid copula type {}'.format(copula_type))
+
+        for subclass in cls.subclasses():
+            if subclass.copula_type is copula_type:
+                return super(Bivariate, cls).__new__(subclass)
+
+    def __init__(self, copula_type=None):
+        """ initialize copula object """
```

Comment:
Please add documentation and example for the init method.  I think it can either be instantiated with a lower case string or a member in CopulaType class?

Reply:
Done.

Not only that, right now we have:
```python
>>> Bivariate(CopulaTypes.FRANK).__class__  # Using CopulaType
copulas.bivariate.frank.Frank

>>> Bivariate('frank').__class__ # Using strings
copulas.bivariate.frank.Frank
```
But thanks to CopulaType being an `enum.Enum`, we can also acces its members by both name and value:

```python
>>> Bivariate(CopulaTypes['FRANK']).__class__
copulas.bivariate.frank.Frank

>>> Bivariate(CopulaTypes(1)).__class__
copulas.bivariate.frank.Frank
```

This is useful in scenarios when we want to instantiate a Copula whose type won't be known until runtime, like when doing the copula selection.